### DISCUSSION TITLE
Constrain noir versions in Nargo.toml

### DISCUSSION THREAD

**Discussion by signorecello** [noir-lang/noir#3140] [Ideas]
### Problem

Different Noir versions produce different results.

So, people cloning repos always have to watch out for the Noir version the repo was built with. And people writing repos always have to say “use Noir 0.xx”

### Happy Case

Having a nargo version property in Nargo.toml that would warn/error if users are using the project with an unsupported version of Nargo.

Example: `noir_version = ^0.16.0` would error if users are using any version above 0.16.x

### Would you like to submit a PR for this Issue?

No

**Comment by TomAFrench** [comment]
For this we need to be able to properly version the language vs `nargo` otherwise we'll be constantly breaking libraries.

**Reply by Savio-Sou** [reply]
Isn't Nargo and the language already being released together under shared version numbers?

**Reply by TomAFrench** [reply]
That's my point. We need to separate these if we want to do this.

**Reply by signorecello** [reply]
Yeah and then use a sort of rust-toolchain file, in this case noir-toolchain

**Reply by Savio-Sou** [reply]
Ah I see.

Makes sense; the more frequent "breaking" doesn't sound that bad though (given we could start with a warning instead of an error) in favour of getting a short-term solution in sooner.

Next steps on this Idea could be:
1. Implement a version mismatch warning per the suggestion based on the current Noir architecture
2. Create a separate Issue for separating Language vs `nargo` versioning
3. Tackle that Issue
4. Create a follow up Issue on turning the version mismatch warning into an error
5. Completes Issue (i.e. completes the full vision)

How does that sound?

**Comment by Savio-Sou** [comment]
Resolved with https://github.com/noir-lang/noir/pull/3336.
